NATIVE-486: Add documentation for CI/CD (#255)
Integreat/integreat-react-native-app: docs/08-cicd.md

Whether `api-triggered` is delivered to production depends on the supplied parameters.

## About the "api-triggered" Workflow 

This workflow can be triggered using the tool `yarn trigger-pipeline` as described [below](#triggering-a-build-in-ci).
Depending on the supplied parameters the build is released to production.

### Triggering a Build in CI

If you decide to do a non-scheduled automatic release of the app the easiest way is to trigger a build in CI/CD.
First you need to get a ["Personal API token"](https://circleci.com/docs/2.0/managing-api-tokens/).
Then you can trigger a build using the tool `yarn trigger-pipeline`. If you execute it without parameters you will see the usage.

# Fastlane

Fastlane is a task-runner for triggering build relevant tasks. It offers integration with XCode and the Android SDK for building and delivering the app.

## Setup of Fastlane

* Install [Ruby >= 2.6.5](https://www.ruby-lang.org/en/documentation/installation/) (The preferred and tested way is to use [Ruby Version Manager (RVM)](https://rvm.io/)).
  * If using RVM you have to do: `rvm use ruby-2.6.5`
* Make sure `ruby --version` reports the correct version

> * Run `bundle install --path vendor/bundle` in the project root, `./android/` and `./ios/`

Reviewer: ```suggestion
* Run `bundle install --path vendor/bundle` in the project root AND in `./android/` AND in `./ios/`
```

## App Store Connect

You can visit the management website for the Play Store [here](https://appstoreconnect.apple.com/). App Store Connect is the product by Apple for managing the App Store presence.

For delivery an [account without 2FA](https://github.com/fastlane/fastlane/blob/b121a96e3e2e0bb83392c130cb3a088c773dbbaf/spaceship/docs/Authentication.md#avoid-2fa-via-additional-account) is required.

### Adding Testers to TestFlight

The [weekly and bi-weekly workflows](#workflows) make the builds directly available to TestFlights "App Store Connect Users". Those should not be confused with "External Tests" which require an approval by apple. Therefore, we currently only use "App Store Connect Users" as testers.

In order to add someone as "App Store Connect User" you have to add the Apple Account to App Store Connect and to TestFlight. This is a two-step process.

### Metadata

The CI/CD pipeline uploads and overwrites metadata during the delivery step.
You can read more about managing metadata for iOS [here](https://docs.fastlane.tools/actions/deliver/).

### Authenticating

> Authentication happens by setting the `FASTLANE_USER` and `FASTLANE_PASSWORD` environment variables as documented [above](#ios-variables). For more information visit the documentation [here](https://docs.fastlane.tools/best-practices/continuous-integration/#environment-variables-to-set) and read about [2-step-verification here](https://github.com/fastlane/fastlane/blob/ee904cd332ab38ca7c1979f0ab58f9389a51fb2d/spaceship/README.md#2-step-verification).

Reviewer: Is this only the case for ios?

Author: yes

## BrowserStack

We are using BrowserStack to run our E2E tests on real iOS and Android devices.

> The general documentation about [E2E tests and BrowserStack can be found here](04-e2e.md). 

Reviewer: ```suggestion
The general documentation about E2E tests and BrowserStack can be found [here](04-e2e.md). 
```

# Continuous Integration and Delivery

# Workflows

Several workflows exist for different purposes. See the table below for an overview.

> |Workflow|Type-checking|Jest tests|Production build|E2E tests|Delivered in production|

Reviewer: The term `Production build` suggests there is also such a thing as `Debug build`. Instead, the only difference is setting the environment variable `E2E_TESTING_IDS`.

Reviewer: ```suggestion
|Workflow|Type-checking|Jest tests|Production build|E2E tests|Delivered to production|
```

|---|---|---|---|---|---|
|commit|:heavy_check_mark:|:heavy_check_mark:|:x:|:heavy_check_mark:|:x:|
|weekly|:heavy_check_mark:|:heavy_check_mark:|:heavy_check_mark:|:x:|:x:|
|bi-weekly|:heavy_check_mark:|:heavy_check_mark:|:heavy_check_mark:|:x:|:heavy_check_mark:|
|api-triggered|:heavy_check_mark:|:heavy_check_mark:|:heavy_check_mark:|:x:|-|

Whether `api-triggered` is delivered to production depends on the supplied parameters.

## About the "api-triggered" Workflow 

This workflow can be triggered using the tool `yarn trigger-pipeline` as described [below](#triggering-a-build-in-ci).

> Depending on the supplied parameters a weekly or bi-weekly build is triggered.

Reviewer: This is not correct. There is no such thing as weekly and bi-weekly builds.

### Triggering a Build in CI

If you decide to do a non-scheduled automatic release of the app the easiest way is to trigger a build in CI/CD.

> First you need to get acquire a ["Personal API token"](https://circleci.com/docs/2.0/managing-api-tokens/).

Reviewer: ```suggestion
First you need to get a ["Personal API token"](https://circleci.com/docs/2.0/managing-api-tokens/).
```

> Then you can trigger a build using the tool `yarn trigger-pipeline`. If you execute it without parameters you will see the usage.

Reviewer: Is there a documentation for this script? If so, this and the lines above should be moved to the corresponding section.

> # Fastlane

Reviewer: When do I need to execute these commands?

Fastlane is a task-runner for triggering build relevant tasks. It offers integration with XCode and the Android SDK for building and delivering the app.

## Setup of Fastlane

> * [Install Ruby >= 2.6.5](https://www.ruby-lang.org/en/documentation/installation/) (The preferred and tested way is using [Ruby Version Manager (RVM)](https://rvm.io/))

Reviewer: ```suggestion
* Install [Ruby >= 2.6.5](https://www.ruby-lang.org/en/documentation/installation/) (The preferred and tested way is to use [Ruby Version Manager (RVM)](https://rvm.io/)).
```

  * If using RVM you have to do: `rvm use ruby-2.6.5`
* Make sure `ruby --version` reports the correct version

> * Run `bundle install --path vendor/bundle` in the project root, `./android/` and `./ios/`

Reviewer: Is it really required to run this command thrice?

* Run `bundle exec fastlane --version`

*Hint: You can use `export FASTLANE_SKIP_UPDATE_CHECK=true` to skip the changelog output*

## Lanes

There are Android and iOS specific lanes. Lanes for Android live in `./android/fastlane` and for iOS in `./ios/fastlane`.

### General 

For an overview of the Fastlane lanes refer to the [auto-generated README](../fastlane/README.md#available-actions). These lanes are responsible for delivering and uploading artifacts, bumping the version using the [deliverino GitHub bot](#deliverino-github) and sending notifications.

### Android

For an overview of the Fastlane lanes refer to the [auto-generated README](../android/fastlane/README.md#available-actions). These lanes are responsible for setting up the signing keys and building the Android app.

### iOS

> For an overview of the Fastlane lanes refer to the [auto-generated README](../ios/fastlane/README.md#available-actions). These lanes are responsible for setting up the certificates and building the iOS app.

Reviewer: Duplicates

# Setup Certificates for iOS on Mac

> The setup of certificates is demonstrated in section [Manual builds](06-manual-builds.md#manually-building-for-ios).

Reviewer: If sections are used in multiple docs, they should probably be moved to a separate file. I don't see an order why some stuff is explained in manual builds and some stuff in cicd.

Author: I moved this to a separate file

## What Types of Certificates Exist?

The certificate system of Apple seems very complex at first. Therefore, we will briefly explain the types of certificates and what the purpose is.

|Certificate Type|Usable for Development|Usable for Testing|Usable for Production|
|---|---|---|---|
|Development|:heavy_check_mark:|:x:|:x:|
|Ad-hoc|:x:|:heavy_check_mark:|:x:|
|Enterprise|:x:|:heavy_check_mark:|:heavy_check_mark:|
|App Store|:x:|:x:|:heavy_check_mark:|

Also, there are various ways to install an iOS app:

* Via a URL which is opened in Safari
* Via App Store

> * Via XCode

Reviewer: Does not really relate to available certificates

> Development certificates are for developers who want to install the app on their devices. You have to select the devices in the Provisioning Profile at [developer.apple.com](https://developer.apple.com/account/resources/profiles/list) you want to install the app on. This means the Device ID (UDID) has to be whitelisted.

Reviewer: ```suggestion
Development certificates are for developers who want to install the app on their devices. You have to select the devices in the [Provisioning Profile](https://developer.apple.com/account/resources/profiles/list) you want to install the app on. This means the Device ID (UDID) has to be whitelisted.
```

Ad-hoc certificates are for public tests with a small amount of users. The difference is that you can provide the testers a URL which they can open on their device to install it. The Device ID (UDID) still has to be whitelisted.

Enterprise certificates allow you to deliver you app within your company or have a bigger public test. It is not possible for small organisation to acquire such a certificate.

App Store certificates allows the upload of the app into the App Store. It is not possible to install a app store signed app directly on your device. It has to be uploaded to App Store Connect.

# Setup Java Keystore for Android

The setup of the JKS is demonstrated in section [Manual builds](06-manual-builds.md#manually-building-for-android).

# Determine the Next Version

The next version of the app must be determined programmatically. The tool `yarn bump-version` can be used. If you execute it without parameters you will see the usage.

> As the version number is based on the release date, the first versions of 2020 are `2020.1.0`, `2020.1.1`, `2020.1.2`...
> 
> The first release february will have the version number `2020.2.0`. 

Reviewer: ```suggestion
The following schema is used for versioning: `YYYY.mm.<release_number>`, where `<release_number>` is a counter for the number of releases in the corresponding month. E.g. the first release in January 2020 uses `2020.01.1`, the second `2020.01.2`.
```

Author: We start at 0. I modified the docu here.

 

# Environment Variables and Dependencies

|Variable|Description|Where do I get it from?|Example|Reference|
|---|---|---|---|---|

> |BROWSERSTACK_ACCESS_KEY|Access Key for BrowserStack|Password Manager|steffen|[Appium REST API](https://www.browserstack.com/app-automate/rest-api)|

Reviewer: I don't think `steffen` is an access key :thinking:
